Autopilot/Simulation/SendToFlightGear/main.cpp

```diff
@@ -0,0 +1,144 @@
+
+    if (packets == NULL)
+    {
+        std::cerr << "System Out of memory, failed to create packet array" << std::endl;
+        return -1;
+    }
+
+    for(int i = 0; i < numDataPoints; i++)
+    {
+        packets[i] = new char[100];
+
+        if (packets[i] == NULL)
+        {
+            std::cerr << "System Out of memory, failed to create packet array" << std::endl;
+            return -1;
+        }
+    }
+
+    fillPackets(packets, numDataPoints);
+
+    sendPacketsToFlightGear(packets, numDataPoints);
+
+    return 0;
+}
+
+static int getNumDatapoints(void)
+{
+    std::ifstream dataFile;
+    std::string line, previousLine;
+    std::string returnChar ("\n");
+    int i = -1;
+
+    dataFile.open("SimulationBuild/SensorOutputs/roll.txt");
+
+    do
+    {
+        getline(dataFile, line);
+        i++;
+    }
+    while (!line.empty());
+
+    dataFile.close();
+
+    return i;
+}
+
+static void fillPackets(char** packets, int numPackets)
+{
+    std::ifstream rollFile;
+    std::string rollLine;
+    std::ifstream pitchFile;
+    std::string pitchLine;
+    std::ifstream yawFile;
+    std::string yawLine;
+
+    std::ifstream latFile;
+    std::string latLine;
+    std::ifstream longFile;
+    std::string longLine;
+    std::ifstream altitudeFile;
+    std::string altitudeLine;
+
+    rollFile.open("SimulationBuild/SensorOutputs/roll.txt");
+    pitchFile.open("SimulationBuild/SensorOutputs/pitch.txt");
+    yawFile.open("SimulationBuild/SensorOutputs/yaw.txt");
+    latFile.open("SimulationBuild/SensorOutputs/latitude.txt");
+    longFile.open("SimulationBuild/SensorOutputs/longitude.txt");
+    altitudeFile.open("SimulationBuild/SensorOutputs/altitude.txt");
+
+    for(int i = 0; i < numPackets; i++)
+    {
+        getline(rollFile, rollLine);
+        getline(pitchFile, pitchLine);
+        getline(yawFile, yawLine);
+        getline(latFile, latLine);
+        getline(longFile, longLine);
+        getline(altitudeFile, altitudeLine);
+
+        // This packet format was defined inside the flight gear protocol file: Simulink-Sim/FlightGear/WargFGPacketStructure.xml If you haven't yet, copy that file into FlightGear's protocols folder
+        sprintf(packets[i], "%.2f,%.6f,%.6f,%.2f,%.2f,%.2f,\n", std::stof(altitudeLine), std::stof(latLine), std::stof(longLine), std::stof(rollLine), std::stof(pitchLine), std::stof(yawLine));
+
+    }
+
+    rollFile.close();
+    pitchFile.close();
+    yawFile.close();
+    latFile.close();
+    longFile.close();
+    altitudeFile.close();
+}
+
+static void sendPacketsToFlightGear(char** packets, int numPackets)
+{
+    QUdpSocket *socket = new QUdpSocket();
+    QHostAddress ipaddr = QHostAddress::LocalHost;
+
+    for (int i = 0; i < numPackets; i++)
+    {
+        socket->writeDatagram(QByteArray((char*)packets[i], strlen(packets[i])), ipaddr, FLIGHTGEAR_UDP_PORT);
+
+        std::this_thread::sleep_for(std::chrono::milliseconds(TIME_BETWEEN_PACKETS_MS));
+    }
+}
```

Comment:
I'll be honest I have no clue what this file does, some comments would be nice so that if someone in the future sees this they can understand what's going on 😅

Reply:
Once the Simulation is run, and the results are produced, the program that this file compiles to packages the data into a format that flightGear understands and sends it off via a series of UDP packets. FLightGear runs on the same machine in my case, so it just sends them off to a udp port that belongs to the localHost ip address.

I think this explanation belongs in confluence docs on the Simulation architecture rather than in comments. (Although I could deff add a comment that links to those docs).